source/common/grpc/grpc_web_filter.cc

```diff
@@ -0,0 +1,109 @@
+#include "common/grpc/grpc_web_filter.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/base64.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Grpc {
+
+const uint8_t GrpcWebFilter::GRPC_WEB_TRAILER = 0b10000000;
+
+GrpcWebFilter::GrpcWebFilter() : is_text_request_(false), is_text_response_(false) {}
+
+GrpcWebFilter::~GrpcWebFilter() {}
+
+// Implements StreamDecoderFilter.
+Http::FilterHeadersStatus GrpcWebFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == content_type->value().c_str()) {
+    is_text_request_ = true;
+  }
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Grpc);
+
+  const Http::HeaderEntry* accept = headers.get(Http::Headers::get().Accept);
+  if (accept != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == accept->value().c_str()) {
+    is_text_response_ = true;
+  }
+  headers.addStatic(Http::Headers::get().TE, Http::Headers::get().TEValues.Trailers);
+  headers.addStatic(Http::Headers::get().GrpcAcceptEncoding,
+                    Http::Headers::get().GrpcAcceptEncodingValues.Default);
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::decodeData(Buffer::Instance& data, bool) {
+  if (!is_text_request_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Parse application/grpc-web-text format.
+  if (data.length() + decoding_buffer_.length() < 4) {
+    decoding_buffer_.move(data);
+    return Http::FilterDataStatus::Continue;
```

Comment:
I cannot use StopIterationAndBuffer, as I have to consume the buffer given by the filter completely and put back b64 decoded data in the same buffer. If I let the filter to buffer the data, I will lose the information about the boundary for the b64 decoded data and raw data in the filter buffer.